ribbon-eureka/src/main/java/com/netflix/loadbalancer/EurekaDynamicNotificationLoadBalancer.java

```diff
@@ -0,0 +1,195 @@
+ *
+ * @author David Liu
+ */
+public class EurekaDynamicNotificationLoadBalancer extends AbstractDynamicServerListLoadBalancer<DiscoveryEnabledServer> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EurekaDynamicNotificationLoadBalancer.class);
+
+    private static class LazyHolder {
+        private static final ExecutorService DEFAULT_SERVER_LIST_UPDATE_EXECUTOR = Executors.newCachedThreadPool(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("EurekaDynamicNotificationLoadBalancer-ServerListUpdate-%d")
+                        .setDaemon(true)
+                        .build()
+        );
+
+        private static final Thread SHUTDOWN_THREAD = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                LOGGER.info("Shutting down the Executor for EurekaDynamicNotificationLoadBalancer");
+                try {
+                    DEFAULT_SERVER_LIST_UPDATE_EXECUTOR.shutdown();
+                    Runtime.getRuntime().removeShutdownHook(SHUTDOWN_THREAD);
+                } catch (Exception e) {
+                    // this can happen in the middle of a real shutdown, and that's ok.
+                }
+
+            }
+        });
+
+        static {
+            Runtime.getRuntime().addShutdownHook(SHUTDOWN_THREAD);
+        }
+    }
+
+    public static ExecutorService getDefaultServerListUpdateExecutor() {
+        return LazyHolder.DEFAULT_SERVER_LIST_UPDATE_EXECUTOR;
+    }
+
+    private final EurekaEventListener updateListener = new EurekaEventListener() {
+        @Override
+        public void onEvent(EurekaEvent event) {
+            if (event instanceof CacheRefreshedEvent) {
+                serverListUpdateService.submit(new Runnable() {  // fire and forget
+                    @Override
+                    public void run() {
+                        EurekaDynamicNotificationLoadBalancer.super.updateListOfServers();
+                    }
+                });
+            }
+        }
+    };
+
+    private final Provider<EurekaClient> eurekaClientProvider;
+    private final ExecutorService serverListUpdateService;
+
+    // necessary for legacy use cases :(
+    // callers of this must ensure initWithNiwsConfig is called also
+    public EurekaDynamicNotificationLoadBalancer() {
+        this(
+                null,
+                new LegacyEurekaClientProvider(),
+                getDefaultServerListUpdateExecutor()
+        );
+    }
+
+    public EurekaDynamicNotificationLoadBalancer(final IClientConfig clientConfig,
+                                                 final Provider<EurekaClient> eurekaClientProvider,
+                                                 ExecutorService serverListUpdateService) {
+        super();
+
+        // wrap around a singleton provider as provider interface does not require this
+        this.eurekaClientProvider = new Provider<EurekaClient>() {
```

Comment:
This seems unnecessary as EurekaClient should be bound as a singleton.  LegacyEurekaClientProvider already takes care of making this singleton.